### PR TITLE
Refactor organelle tooltips

### LINKED ISSUE #1809: Rework (editor) tooltips to make translations possible to make properly
Right now many things in the tooltips are translated one word at a time (especially in the organelle tooltips).
Instead the entire thing should be translatable at once. Might be a bit difficult to do, we might need a custom rich text parser to allow the image etc. elements to be positioned around the text how they need to be.

### LINKED ISSUE #1980: Rewrite SelectionMenuToolTip.WriteOrganelleProcessList to use ChemicalEquation
in order to reduce code duplication. This will likely require a few changes in ChemicalEquation to allow the process name to be changed (and center aligned).
Making numbers red if they are 0 can be already done with the mark red compound list.

### PR DESCRIPTION
**Pull request by Kasterisk**
Improves organelle tooltips so translating the processes descriptions could be done properly. Compound names and keybindings are now embedded as "keys" on the strings which then could be parsed and extracted in the code to be replaced with actual values. Suggestions welcome. Partially addresses #1913.

Also made some internal code refactoring. 

Fixes #1809
Closes #1980

### CODE REVIEW

File: src/gui_common/fonts/BBCode-Image-VerticalCenterAlign-3.tres [hunk 1]
```diff
@@ -0,0 +1,4 @@
+[gd_resource type="BitmapFont" format=2]
```
Comment (Kasterisk) [inline src/gui_common/fonts/BBCode-Image-VerticalCenterAlign-3.tres:1]: This is needed to properly center align images vertically on rich text label: https://docs.godotengine.org/en/stable/tutorials/gui/bbcode_in_richtextlabel.html#image-vertical-offset

File: src/general/MainMenu.cs [hunk 2]
```diff
@@ -191,7 +191,9 @@ private void SwitchMenu()
 
     private void OnIntroEnded()
     {
-        TransitionManager.Instance.AddScreenFade(ScreenFade.FadeType.FadeOut, 0.5f, false);
+        var duration = IsReturningToMenu ? 0.3f : 0.5f;
+
+        TransitionManager.Instance.AddScreenFade(ScreenFade.FadeType.FadeOut, duration, false);
```
Comment (hhyyrylainen) [inline src/general/MainMenu.cs:196]: I would just put the ternary operator right in this call

File: src/gui_common/ChemicalEquation.cs [hunk 3]
```diff
@@ -84,6 +85,22 @@ public bool ShowSpinner
 
     public float SpinnerBaseSpeed { get; set; } = Constants.DEFAULT_PROCESS_SPINNER_SPEED;
 
+    public Color DefaultTitleColour
+    {
+        get => defaultTitleColour;
+        set
+        {
+            defaultTitleColour = value;
```
Comment (hhyyrylainen) [inline src/gui_common/ChemicalEquation.cs:93]: I'd add an early return if `value == defaultTitleColour` here

File: src/gui_common/ChemicalEquation.cs [hunk 4]
```diff
@@ -92,7 +109,6 @@ public override void _Ready()
 
         equationArrowTexture = GD.Load<Texture>("res://assets/textures/gui/bevel/WhiteArrow.png");
 
-        spinner.Visible = showSpinner;
```
Comment (hhyyrylainen) [inline src/gui_common/ChemicalEquation.cs:95]: Why was this removed?

Comment (hhyyrylainen) [reply src/gui_common/ChemicalEquation.cs:95]: It seems now that the method that now sets spinner.Visible isn't called here anymore? Does the process panel display still work?

File: src/gui_common/ChemicalEquation.cs [hunk 5]
```diff
@@ -162,6 +178,19 @@ private void UpdateEquation()
...
+    {
+        spinner.Visible = ShowSpinner;
+        title.Text = EquationFromProcess.Name;
+
+        var colour = DefaultTitleColour;
+
+        if (MarkRedOnLimitingCompounds && EquationFromProcess.LimitingCompounds.Count > 0)
+            colour = new Color(1.0f, 0.3f, 0.3f);
+
+        title.AddColorOverride("font_color", colour);
```
Comment (hhyyrylainen) [inline src/gui_common/ChemicalEquation.cs:191]:
```suggestion
        if (MarkRedOnLimitingCompounds && EquationFromProcess.LimitingCompounds.Count > 0){
            title.AddColorOverride("font_color", new Color(1.0f, 0.3f, 0.3f));
        } else {
            title.AddColorOverride("font_color", DefaultTitleColour);
        }
```
(but formatter correctly)

File: src/gui_common/ChemicalEquation.cs [hunk 6]
```diff
@@ -22,6 +22,7 @@ public class ChemicalEquation : VBoxContainer
     private IProcessDisplayInfo equationFromProcess;
     private bool showSpinner;
     private Texture equationArrowTexture;
+    private Color defaultTitleColour;
```
Comment (hhyyrylainen) [inline src/gui_common/ChemicalEquation.cs:25]: Shouldn't this default to white?

File: src/microbe_stage/ProcessSpeedInformation.cs [hunk 7]
```diff
@@ -1,48 +1,42 @@
...
+public class ProcessSpeedInformation : IProcessDisplayInfo
 {
+    public readonly Dictionary<Compound, float> WritableInputs = new Dictionary<Compound, float>();
+    public readonly Dictionary<Compound, float> WritableOutputs = new Dictionary<Compound, float>();
+    public readonly List<Compound> WritableLimitingCompounds = new List<Compound>();
+
+    public readonly Dictionary<Compound, float> AvailableAmounts = new Dictionary<Compound, float>();
+
+    // ReSharper disable once CollectionNeverQueried.Global
+    public readonly Dictionary<Compound, float> AvailableRates = new Dictionary<Compound, float>();
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSpeedInformation.cs:17]: these should be get only properties, right?

File: src/microbe_stage/ProcessSystem.cs [hunk 8]
```diff
@@ -244,18 +244,16 @@ private static float GetDissolvedInBiome(Compound compound, BiomeConditions biom
...
 
-            input.AvailableAmount = availableInEnvironment;
+            result.AvailableAmounts[entry.Key] = availableInEnvironment;
 
             // More than needed environment value boosts the effectiveness
-            input.AvailableRate = availableInEnvironment / entry.Value;
+            result.AvailableRates[entry.Key] = availableInEnvironment / entry.Value;
 
-            speedFactor *= input.AvailableRate;
+            speedFactor *= result.AvailableRates[entry.Key];
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSystem.cs:254]: unnecessary read back from the dictionary `availableInEnvironment / entry.Value` should be stored in a variable so it can be used in multiple places here.

File: src/microbe_stage/ProcessSpeedInformation.cs [hunk 9]
```diff
@@ -1,48 +1,42 @@
...
+        AvailableAmounts.Where(p => p.Key.IsEnvironmental);
 
-    public class EnvironmentalInput : CompoundAmount
-    {
-        public EnvironmentalInput(Compound compound, float amount)
-            : base(compound, amount)
-        {
-        }
+    // ReSharper disable once UnassignedGetOnlyAutoProperty
+    public IReadOnlyDictionary<Compound, float> FullSpeedRequiredEnvironmentalInputs { get; }
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessSpeedInformation.cs:35]: Why is this not assigned, isn't this then always null?

File: src/microbe_stage/ProcessList.cs [hunk 10]
```diff
@@ -12,6 +12,19 @@ public class ProcessList : VBoxContainer
 
     public List<IProcessDisplayInfo> ProcessesToShow { get; set; }
 
+    public bool ShowSpinners { get; set; } = true;
+
+    /// <summary>
+    ///   The default color for all the processes title in this list.
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessList.cs:18]:
```suggestion
    ///   The default color for all the process titles in this list.
```

File: src/microbe_stage/ProcessList.cs [hunk 11]
```diff
@@ -12,6 +12,19 @@ public class ProcessList : VBoxContainer
...
+    /// </summary>
+    public Color ProcessesTitleColour { get; set; } = Colors.White;
+
+    /// <summary>
+    ///   Controls whether the color of all processes title in this list should be turned
```
Comment (hhyyrylainen) [inline src/microbe_stage/ProcessList.cs:23]: I think this comment is unclear, does this mean:
- The individual process name turns red if it has limiting compounds
- Or all of the processes turn red if any has limiting compounds
?

File: src/microbe_stage/Compound.cs [hunk 12]
```diff
@@ -72,6 +79,14 @@ public void Check(string name)
         TranslationHelper.CopyTranslateTemplatesToTranslateSource(this);
     }
 
+    public void Resolve()
+    {
+        if (!string.IsNullOrEmpty(IconPath))
```
Comment (hhyyrylainen) [inline src/microbe_stage/Compound.cs:84]: I think you should add a check against the Iconpath being null or empty in the Check method, around line 53

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 13]
```diff
@@ -68,6 +72,29 @@ public string DisplayName
         }
     }
 
+    /// <summary>
+    ///   Description of processes an organelle does if any.
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     This supports custom format (for example: "Turns [glucose] into [atp]") where strings inside
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: Would it make sense to namespace the bbcode, like: `[compound:glucose]` that way we could have a complex setup of  bbcode parsers (for example the thrivepedia might find this useful) attached to one rich text label?

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 14]
```diff
@@ -68,6 +72,29 @@ public string DisplayName
...
+    ///     the square brackets will be parsed and replaced with a predefined template. This is done to
+    ///     make translating feasible.
+    ///     NOTE: description string should only be set here and not directly on the rich text label node
+    ///     as it will be overidden otherwise.
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:84]: Can this note be added as an editor description on this class' scene as well?

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 15]
```diff
@@ -68,6 +72,29 @@ public string DisplayName
...
+    ///   </para>
+    /// </remarks>
+    [Export]
+    public string ProcessesDescription
+    {
+        get => processesDescription;
+        set
+        {
+            processesDescription = value;
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:93]: I think checking if the new value is the same as old would be worth it here to avoid expensive operation if it the text didn't change.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 16]
```diff
@@ -99,20 +137,41 @@ public override void _Ready()
...
+    public override void _Notification(int what)
+    {
+        if (what == NotificationTranslationChanged)
+        {
+            UpdateProcessesDescription();
+        }
+
+        if (what == NotificationResized)
+        {
+            // A workaround to get RichTextLabel's height properly updated on size change
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:163]: Is this a known bug in Godot? Should we open an issue in the thrive repo and link to that here about improving this?
I imagine we'll use rich text elsewhere as well and that might also need this workaround...

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 17]
```diff
@@ -335,6 +278,52 @@ public void OnHide()
...
+    ///   "keys" and turns them into a BBCode string with matching values.
+    /// </summary>
+    private string ParseProcessesDescription()
+    {
+        if (string.IsNullOrEmpty(ProcessesDescription))
+            return string.Empty;
+
+        var inputEvents = Settings.GetCurrentlyAppliedControls();
+
+        var result = Regex.Replace(TranslationServer.Translate(ProcessesDescription), @"\[(.*?)\]", found =>
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:292]: I think this part (see one of my previous comments) would be worth it to generalize. Also there should be an escape option `\[stuff]` should be output as `[stuff]` and not attempted to be parsed.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 18]
```diff
@@ -335,6 +278,52 @@ public void OnHide()
...
+            }
+
+            // Parse input actions
+            if (InputMap.HasAction(value))
+            {
+                var events = inputEvents.Data[value];
+
+                for (int i = 0; i < events.Count; i++)
+                {
+                    parsed += $"[b][{KeyNames.Translate(events[i].Code)}][/b]";
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:314]: I think this could share some logic with the key button prompts, as these really should share the same logic to determine what to show in the GUI text and on the button press prompts.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs [hunk 19]
```diff
@@ -335,6 +278,52 @@ public void OnHide()
...
+                parsed = $"[b]{compound.Name}[/b] [font=res://src/gui_common/fonts/" +
+                    $"BBCode-Image-VerticalCenterAlign-3.tres] [img=20]{compound.IconPath}[/img][/font]";
+            }
+
+            // Parse input actions
+            if (InputMap.HasAction(value))
+            {
+                var events = inputEvents.Data[value];
+
+                for (int i = 0; i < events.Count; i++)
```
Comment (hhyyrylainen) [inline src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:312]: I think only the primary binding should be shown here to reduce clutter in case someone adds like 5 key bindings for an action.

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Overall this approach seems workable, I have some comments regarding the specifics.

File: src/gui_common/ChemicalEquation.cs (hunk 4 above)
Comment (Kasterisk) [reply src/gui_common/ChemicalEquation.cs:95]: It's now set in `UpdateHeader` which will be called inside the `UpdateEquation` call just below this.

Comment (hhyyrylainen) [reply src/gui_common/ChemicalEquation.cs:95]: Ah okay, I just looked for UpdateHeader call here, didn't consider that another method that's called here would call UpdateHeader.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs (hunk 16 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:163]: I think it is, I'm using RichTextLabel's [fit_content_height ](https://docs.godotengine.org/en/stable/classes/class_richtextlabel.html#class-richtextlabel-property-fit-content-height) property which in itself is a workaround as noted in the docs and will be removed in future releases (possibly replaced in 4.0) so I don't think we have to open an issue(?). Enabling the property is kind of necessary, otherwise rich text label could hardly be used in the tooltips...

Comment (hhyyrylainen) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:163]: Well could you open an issue reminding that this should be revisited once Godot 4 comes out and link that new issue here so that we can remember to do that.

Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:163]: Sure opened a new issue #2236

File: src/microbe_stage/ProcessSpeedInformation.cs (hunk 7 above)
Comment (Kasterisk) [reply src/microbe_stage/ProcessSpeedInformation.cs:17]: No these are needed to be able being set in other places, will remove the readonlys.

Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSpeedInformation.cs:17]: At least make them properties with public set, then.

Comment (hhyyrylainen) [reply src/microbe_stage/ProcessSpeedInformation.cs:17]: Actually, hang on. How did this code even work in the first place if these readonlies are set from other places? That can't compile.

Comment (Kasterisk) [reply src/microbe_stage/ProcessSpeedInformation.cs:17]: Wait, now that I think about it, making these get only properties makes sense as the actual assigning currently only happens here (hence the readonlys working). I must have confused the modifying of a dictionary keyvalue pairs with it requiring a setter, woops...

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs (hunk 17 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:292]: I've now moved the logic to separate class that inherits RichTextLabel so now it should be reusable for other places (just drag and drop the rich text label scene, script is already attached). Also added the mentioned escape option.

File: src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs (hunk 13 above)
Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: Want to mention that the format is not really bbcode (in the sense of tags) as the bracketed strings just serves as key so the regex could replace them with an actual bbcode string. That said, I think having a namespace is nice as it would allow different ways of parsing the same key which is very convenient in different contexts, though I feel adding such option could be reserved for later once features like Thrivepedia is implemented as for now the format usage is simple enough (stylizing compound name and inputs for tooltips).

Comment (hhyyrylainen) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: Even now this conflicts with normal bbcode, right? In that if there is a [b] in the text, your code would try to act as if that was an organelle name, right? I think that writing a simple proper parser would be a good idea even now to solve the code trying to work on unknown tags etc. Actually reading that Godot documentation you linked, I was kinda hoping that this would have been implemented as a bbcode extension to Godot, instead of just a transform on the actual text.

Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: Regarding conflict with bbcode tags it's possbile to make it so they will be ignored (returned unchanged by regex replace) since they don't match any compound names or inputs so they'll just work normally, I tried adding a strikethrough tag between one of the keys and it also work. Also it's unlikely a key and a bbcode tag has the same name, if it does I understand the need for a namespace. I'm assuming you're referring to [this](https://docs.godotengine.org/en/stable/tutorials/gui/bbcode_in_richtextlabel.html#custom-bbcode-tags-and-text-effects) part of the documentation, as far as I know it only allows adding custom tags for text effect purposes (which only operates per string characters, can't add images etc) which is why I decided to go with this approach.

Comment (hhyyrylainen) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: They work now because the code tries *any* bbtag to see if it is a valid compound name, and if it is, it is replaced. That's why the default tags still work.
That earlier idea about namespacing the tags would help avoid conflicts with inbuilt names *and* reduce the number of lookups "is this a compound name?" needed.
Yes that was the documentation I was referring to. So I guess this "preprocessor" approach to the bbcode is needed then.
Btw one more thing I noticed about the syntax, it's `[glucose]` which isn't valid bbcode, for consistency with other tags shouldn't we make the parser look for `[compound]glucose[/compound]` ? Or maybe even combining the namespaces: `[thrive:compound]glucose[/thrive:compound]` ?

Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: I feel like `[glucose]` is much simpler, but I'm fine with making it consistent with bbcodes so I'll try do that.

Comment (hhyyrylainen) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: Another idea would be to follow C# string formatting by using `{compound:glucose}` but I'm not sure if the automatic weblate C# template string checking would explode on that...
What I initially had in mind for the tooltips was that `{0}` etc. would be used in the template string and that would be replaced with the right compound names, but with your approach that probably isn't possible (my assumption was that the code doing the template text parsing knew what compound names etc. it would want to insert within the text placeholders).

Comment (Kasterisk) [reply src/gui_common/tooltip/microbe_editor/SelectionMenuToolTip.cs:80]: I'd guess probably yeah lol

> What I initially had in mind for the tooltips was that {0} etc. would be used in the template string and that would be replaced with the right compound names, but with your approach that probably isn't possible (my assumption was that the code doing the template text parsing knew what compound names etc. it would want to insert within the text placeholders).

That's also what I had in mind initially, but then thought that might get a bit messy as there's multiple places to maintain for ordering or something, with this it's much simpler and kind of like "all in one" I think.

**Comment by Kasterisk** [comment]
Alright, I've switched to a bbcode-styled tags for the format along with thrive namespace so the custom parser will not try to parse default tags potentially saving performance. Let me know if anything is not right.

**Comment by Kasterisk** [comment]
I think I've managed to get it close to what you're describing. Works nicely with default bbcodes so far.

File: src/gui_common/CustomRichTextLabel.cs [hunk 20]
```diff
@@ -0,0 +1,245 @@
...
+        {
+            case ThriveBbCodeTag.Compound:
+            {
+                if (SimulationParameters.Instance.DoesCompoundExist(input))
+                {
+                    var compound = SimulationParameters.Instance.GetCompound(input);
+
+                    output = $"[b]{compound.Name}[/b] [font=res://src/gui_common/fonts/" +
+                        $"BBCode-Image-VerticalCenterAlign-3.tres] [img=20]{compound.IconPath}[/img][/font]";
+                }
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:226]: I think printing a warning if not found would be a good idea to make it more visible that there is some invalid bbcode.

File: src/gui_common/CustomRichTextLabel.cs [hunk 21]
```diff
@@ -0,0 +1,245 @@
...
+                break;
+            }
+
+            case ThriveBbCodeTag.Input:
+            {
+                if (InputMap.HasAction(input))
+                {
+                    output = "[font=res://src/gui_common/fonts/BBCode-Image-VerticalCenterAlign-9.tres]" +
+                        $"[img=30]{KeyPromptHelper.GetPathForAction(input)}[/img][/font]";
+                }
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:237]: and here as well

File: src/gui_common/CustomRichTextLabel.cs [hunk 22]
```diff
@@ -0,0 +1,245 @@
...
+        var tagStack = new Stack<string>();
+
+        var isIteratingTag = false;
+        var isIteratingContent = false;
+
+        // The index of a closing bracket in a last iterated opening tag, used
+        // to retrieve the tagged substring
+        var lastStartingTagEndIndex = 0;
+
+        for (int index = 0; index < ExtendedBbcode.Length; index++)
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:77]: use preincrement

Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:77]: I think just for safety regarding performance, I'd prefer if this directly read the object field here instead of going through the field (as that incurs an extra method call in C# and we definitely want this parser to be fast)

File: src/gui_common/CustomRichTextLabel.cs [hunk 23]
```diff
@@ -0,0 +1,245 @@
...
+                isIteratingTag = true;
+
+                // Skip once, so the bracket doesn't get added into the final string
+                continue;
+            }
+
+            // Character is not a tag, write it normally into the final string
+            if (!isIteratingTag && !isIteratingContent)
+            {
+                result.Append(character);
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:99]: I'll let this be, but could you add a TODO comment about trying to add entire substrings at once, as at least in C++ that's the faster approach, instead of going character by character.

File: src/gui_common/CustomRichTextLabel.cs [hunk 24]
```diff
@@ -0,0 +1,245 @@
...
+        BbcodeEnabled = true;
+    }
+
+    /// <summary>
+    ///   Parses a custom tagged substring into a templated bbcode.
+    /// </summary>
+    private void ParseCustomTags()
+    {
+        var result = new StringBuilder();
+        var currentTagBlock = new StringBuilder();
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:66]: give this a default capacity of 50

File: src/gui_common/CustomRichTextLabel.cs [hunk 25]
```diff
@@ -0,0 +1,245 @@
...
+        // Make sure bbcode is enabled
+        BbcodeEnabled = true;
+    }
+
+    /// <summary>
+    ///   Parses a custom tagged substring into a templated bbcode.
+    /// </summary>
+    private void ParseCustomTags()
+    {
+        var result = new StringBuilder();
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:65]: Give this a default capacity of the input text (to avoid a few resizings of the internal buffer).

File: src/gui_common/CustomRichTextLabel.cs [hunk 26]
```diff
@@ -0,0 +1,245 @@
...
+            }
+
+            if (isIteratingTag)
+            {
+                // Keep iterating until we hit a closing bracket
+                if (character != ']')
+                {
+                    currentTagBlock.Append(character);
+
+                    // No closing bracket found, just write normally to the final string and abort trying to parse
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:109]: In my code style I add the what the if result is inside the if, so I'd put this after the brace on line 111

File: src/gui_common/CustomRichTextLabel.cs [hunk 27]
```diff
@@ -0,0 +1,245 @@
...
+                        isIteratingTag = false;
+                        continue;
+                    }
+
+                    // Finally try building the bbcode template for the tagged substring
+
+                    var closingTagStartIndex = ExtendedBbcode.Find("[", lastStartingTagEndIndex);
+
+                    var input = ExtendedBbcode.Substr(
+                        lastStartingTagEndIndex + 1, closingTagStartIndex - lastStartingTagEndIndex - 1);
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:170]: Like I'd imagine this similar sub string add would be used instead of character by character append to the result.

File: src/gui_common/CustomRichTextLabel.cs [hunk 28]
```diff
@@ -0,0 +1,245 @@
...
+
+                    ThriveBbCodeTag parsedTag;
+
+                    if (Enum.TryParse(tagStack.Peek(), true, out parsedTag))
+                    {
+                        // Success!
```
Comment (hhyyrylainen) [inline src/gui_common/CustomRichTextLabel.cs:176]: this is the right if comment style

**Review by hhyyrylainen — CHANGES_REQUESTED** [review]
Just a few more minor code comments (though, I haven't playtested this yet so I may notice some issue then).

**Review by hhyyrylainen — APPROVED** [review]
